lib/service/awsoidc.go

```diff
@@ -0,0 +1,245 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/auth/integration/integrationv1"
+	"github.com/gravitational/teleport/lib/automaticupgrades"
+	"github.com/gravitational/teleport/lib/integrations/awsoidc"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils/interval"
+
+	ststypes "github.com/aws/aws-sdk-go-v2/service/sts/types"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"golang.org/x/time/rate"
+)
+
+const (
+	// updateDeployAgentsInterval specifies how frequently to check for available updates.
+	updateDeployAgentsInterval = time.Minute * 30
+
+	// updateDeployAgentsRateLimit specifies the time between updates across AWS regions.
+	updateDeployAgentsRateLimit = time.Second * 30
+)
+
+func (process *TeleportProcess) periodUpdateDeployServiceAgents() error {
+	if !process.Config.Auth.Enabled {
+		return nil
+	}
```

Comment:
I decided to run this in the auth instead of proxy for two reasons

- The proxy service does not have permissions to read the cluster_maintenance_config that contains the upgrade window configuration.
- The proxy services are not configured with the TELEPORT_AUTOMATIC_UPGRADES env variable.

If you feel strongly that this task should run in the proxy we can make the necessary changes. But it is a bit simpler to run the task on auth.

Reply:
I really think we should do this within the proxy.
Auth should be reserved for more critical tasks (eg generating certs and tokens).

Adding read permissions for `cluster_maintenance_config` should be ok.
The `TELEPORT_AUTOMATIC_UPGRADES` should be available in the `.../webapi/ping` endpoint, so I think Proxy should already have access to that value.

What do you think @r0mant ?

Reply:
> Auth should be reserved for more critical tasks (eg generating certs and tokens).

That's a valid concern. I've switch the task to run on proxies.